Log a warning when best model is not loaded

akwako opened the pull request:
Specifying "load_best_model_at_end=True" in TrainingArguments should ensure that the best model is always loaded at end of training. However, this is not always the case: When best_model_checkpoint is None, the best model is not loaded, and the user may be unaware of this behavior.

Add a warning to the log to let the user know when the best model is not loaded at the end of training. Suggest that the user check the "save_strategy" TrainingArguments, as failing to to do so is one possible reason why the best model failed to load.

@muellerzr 
@pacman100

akwako committed:
Log a warning when best model is not loaded

Specifying "load_best_model_at_end=True" in TrainingArguments should
ensure that the best model is always loaded at end of training.
However, this is not always the case: When best_model_checkpoint is
None, the best model is not loaded, and the user may be unaware of this
behavior.

Add a warning to the log to let the user know when the best model is
not loaded at the end of training. Suggest that the user check the
"save_strategy" TrainingArgument, as failing to to do so is one
possible reason why the best model failed to load.

amyeroberts reviewed (commented):
Hi @akwako - thanks for opening this PR and contributing to improving the library! 

What are the circumstances when `args.load_best_model_at_end` is `True` and `self.best_model_checkpoint` is `None` i.e. what other argument settings can cause this? 

Instead of raising a warning at the end of a loop, I'd suggest adding argument verification in the trainer's init. This ensures failing fast.

akwako committed:
Ensure max_steps > save_steps

If max_steps < args.save_steps, and save_strategy="steps", then models
will not be saved. Additionally, state.best_model_checkpoint will be
None at the end training, which means that the best model will not be
loaded at the end of trianing, even if args.load_best_model_at_end=True.

Raise error if max_steps < args.save_steps and save_strategy="steps".
Inform user to change TrainingArguments to avoid issue.

akwako commented:
Hi @amyeroberts, 

Thank you for your reply and feedback!

As far as I can tell, the main issue is when max_steps < save_steps, and only when save_strategy="steps". Since max_steps is computed in the _inner_training_loop, I have added a check there to make sure training fails fast, as you suggested. However, in case this is happening in other ways, it might be good to keep the warning in the log, too. What do you think?

amyeroberts commented on src/transformers/trainer.py:
```diff
@@ -1613,6 +1613,14 @@ def _inner_training_loop(
             )
 
+        # Ensure save_steps < max_steps
+        if self.args.save_strategy == "steps":
+            if max_steps < self.args.save_steps:
+                raise ValueError(
+                    'args.save_steps must be less than max_steps. Either set args.save_strategy to "no" or "epoch", '
+                    'or increase number of training steps.'
+                )
```
```suggestion
        if self.args.save_strategy == "steps" and max_steps < self.args.save_steps:
            raise ValueError(
                'args.save_steps must be less than max_steps. Either set args.save_strategy to "no" or "epoch", '
                'or increase number of training steps.'
            )
```

amyeroberts commented on src/transformers/trainer.py:
```diff
@@ -1968,8 +1976,12 @@ def _inner_training_loop(
+        elif args.load_best_model_at_end and self.state.best_model_checkpoint is None:
+            logger.warning(
+                'Best model cannot be loaded because "trainer.state.best_model_checkpoint" is None. '
+                'Have you set the "save_strategy" Training Argument?'
```
For my own understanding - what values of `"save_strategy"` are expected to work or not work here? If we're to emit a log here - we should be explicit about exactly what the user has to do. Ideally we'd check that save strategy is compatible with `load_best_model_at_end`.

akwako replied:
Values of `save_strategy` that would work are `save_strategy="epoch"` OR, if `save_strategy="steps"`, then `save_steps` should be set lower than `max_steps`.

However, since adding the check in `_inner_training_loop`, the problematic behavior shouldn't be occurring, so it doesn't seem necessary to specify. Perhaps removing the line that reads, 'Have you set the "save_strategy" Training Argument?' Or we can leave out the log warning entirely, if you think that's best.

amyeroberts reviewed (commented):
Thanks for adding the argument validation! 

I think any log that is emitted should be very clear about how the user can fix it and, ideally, not show at the end of training. 

My concern at the moment is this is emitting a warning for a behaviour which shouldn't be happening at all - which is what we want to fix instead of a message. 

cc @pacman100 @muellerzr  for your thoughts

amyeroberts commented:
Possibly related to #27332

akwako committed:
Make style correction

Co-authored-by: amyeroberts <22614925+amyeroberts@users.noreply.github.com>

HuggingFaceDocBuilderDev commented:
The docs for this PR live [here](https://moon-ci-docs.huggingface.co/docs/transformers/pr_28500). All of your documentation changes will be reflected on that endpoint. The docs are available until 30 days after the last update.

akwako commented:
Hi @amyeroberts, 

Thanks for your suggestions! 

I looked into [#27332](https://github.com/huggingface/transformers/issues/27332), and I think you're right. When I run it on my computer, I get the (newly added) message `ValueError: args.save_steps must be less than max_steps. Either set args.save_strategy to "no" or "epoch", or increase number of training steps.`

akwako committed:
Raise error only if load_best_model_at_end=True

Log a warning if model(s) will not be saved during training due to
max_steps being less than args.save_steps. Do not raise a ValueError,
however, unless the user has specified args.load_best_model_at_end=True.

Model will not be saved during training if max_steps < args.save_steps.
This may not be an issue, however, unless the user is explictly asking
for model saves; furthermore, raising a ValueError ensuring that
max_steps > args.save_steps leads to failed pytorch tests / checks.
Instead, raise a ValueError only in the case that the user asks that
the best model be loaded at end of training.

akwako committed:
Amend logger warning

akwako committed:
Make style corrections
